### PR TITLE
Misc Bebop updates

### PR DESCRIPTION
**Pull request by alexdewar**
Hi,

I've just been tidying up some of the (many) old branches we have and I figured these various Bebop bits could be merged.

Commit summary:
* Add support for Bebop's relative move command (i.e. you can set a pose relative to the current position)
* Add initial work on UAV navigation code

I hit a bit of a block with the UAV navigation stuff as the Bebop kept drifting around too much, so for now there is just support for setting and holding a pose. I was thinking this could be a useful base for when we get support for the new drone added and then have a PID controller to counteract the drifting.

### CODE REVIEW

File: libbebop/bebop.cc [hunk 1]
```diff
@@ -235,6 +236,36 @@ Bebop::doFlatTrimCalibration()
     m_FlatTrimSemaphore.wait();
 }
 
+//! Move to a new pose relative to the current one
+void
+Bebop::relativeMove(meter_t dx, meter_t dy, meter_t dz, radian_t dpsi)
```
Comment (neworderofjamie) [inline libbebop/bebop.cc]: What's a ``dpsi`` angle?

File: libbebop/bebop.h [hunk 2]
```diff
@@ -283,6 +301,9 @@ class Bebop
     LimitValues<meters_per_second_t> m_VerticalSpeedLimits;
     LimitValues<degrees_per_second_t> m_YawSpeedLimits;
     std::atomic<unsigned char> m_BatteryLevel;
+    std::atomic<RelativeMoveState> m_RelativeMoveState{ RelativeMoveState::Success };
```
Comment (neworderofjamie) [inline libbebop/bebop.h]: Is there a reason this isn't initialised in the err 'initial' state? :smile:

File: libbebop/bebop.cc (hunk 1 above)
Comment (alexdewar) [reply libbebop/bebop.cc]: Change in yaw -- it's what the Bebop API calls it. ``yaw`` would probably be a better name for this though.

File: libbebop/bebop.h (hunk 2 above)
Comment (alexdewar) [reply libbebop/bebop.h]: Hmm it doesn't *look* like there's a good reason... but if there was, I wrote this so long ago I wouldn't remember anyway :wink: . I'll fix this up

**Comment by alexdewar** [comment]
Hmm... I wasn't including ``bebop_relative_move_test`` in the build before, and now that I do it seems not to compile on your machine.

```
g++ -o bebop_relative_move_test bebop_relative_move_test.cc -march=native -std=c++14 -Wall -Wpedantic -Wextra -MMD -MP -I/home/jenkins/workspace/B_robotics_bob_robotics_uav_misc/make_common/.. -O2 -pipe -DDISABLE_PREDEFINED_UNITS -DENABLE_PREDEFINED_LENGTH_UNITS -DENABLE_PREDEFINED_TIME_UNITS -DENABLE_PREDEFINED_ANGLE_UNITS -DENABLE_PREDEFINED_VELOCITY_UNITS -DENABLE_PREDEFINED_ANGULAR_VELOCITY_UNITS `pkg-config --cflags opencv` -I/opt/arsdk/out/arsdk-native/staging/usr/include -Wl,-rpath,/opt/arsdk/out/arsdk-native/staging/usr/lib,--disable-new-dtags -Wno-implicit-fallthrough -lm -lstdc++ -pthread `pkg-config --libs opencv` -L/home/jenkins/workspace/B_robotics_bob_robotics_uav_misc/make_common/../libbebop -lbebop /opt/arsdk/out/arsdk-native/staging/usr/lib/libarsal.so /opt/arsdk/out/arsdk-native/staging/usr/lib/libardiscovery.so /opt/arsdk/out/arsdk-native/staging/usr/lib/libarcontroller.so /opt/arsdk/out/arsdk-native/staging/usr/lib/libarnetworkal.so /opt/arsdk/out/arsdk-native/staging/usr/lib/libarcommands.so /opt/arsdk/out/arsdk-native/staging/usr/lib/libmux.so /opt/arsdk/out/arsdk-native/staging/usr/lib/libpomp.so /opt/arsdk/out/arsdk-native/staging/usr/lib/libjson-c.so.2 /opt/arsdk/out/arsdk-native/staging/usr/lib/libarstream.so /opt/arsdk/out/arsdk-native/staging/usr/lib/libarstream2.so /opt/arsdk/out/arsdk-native/staging/usr/lib/libarnetwork.so /opt/arsdk/out/arsdk-native/staging/usr/lib/librtsp.so /opt/arsdk/out/arsdk-native/staging/usr/lib/libsdp.so /opt/arsdk/out/arsdk-native/staging/usr/lib/libulog.so /opt/arsdk/out/arsdk-native/staging/usr/lib/libarmedia.so /opt/arsdk/out/arsdk-native/staging/usr/lib/libfutils.so -lavcodec -lavformat -lavutil -lswscale
/home/jenkins/workspace/B_robotics_bob_robotics_uav_misc/make_common/../libbebop/libbebop.a(video.o): In function `BoBRobotics::Robots::Bebop::VideoStream::readFrame(cv::Mat&)':
video.cc:(.text+0x184): undefined reference to `cv::Mat::copyTo(cv::_OutputArray const&) const'
/home/jenkins/workspace/B_robotics_bob_robotics_uav_misc/make_common/../libbebop/libbebop.a(video.o): In function `BoBRobotics::Robots::Bebop::VideoStream::~VideoStream()':
video.cc:(.text+0x5df): undefined reference to `cv::fastFree(void*)'
video.cc:(.text+0x66d): undefined reference to `cv::Mat::deallocate()'
video.cc:(.text+0x67d): undefined reference to `cv::Mat::deallocate()'
/home/jenkins/workspace/B_robotics_bob_robotics_uav_misc/make_common/../libbebop/libbebop.a(video.o): In function `BoBRobotics::Robots::Bebop::VideoStream::frameCallback(ARCONTROLLER_Frame_t*, void*)':
video.cc:(.text+0x18f5): undefined reference to `cv::Mat::updateContinuityFlag()'
video.cc:(.text+0x19e8): undefined reference to `cv::fastFree(void*)'
video.cc:(.text+0x1aab): undefined reference to `cv::fastFree(void*)'
video.cc:(.text+0x1aed): undefined reference to `cv::Mat::deallocate()'
video.cc:(.text+0x1b4f): undefined reference to `cv::error(int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const*, char const*, int)'
/home/jenkins/workspace/B_robotics_bob_robotics_uav_misc/make_common/../libbebop/libbebop.a(video.o): In function `BoBRobotics::Robots::Bebop::VideoStream::~VideoStream()':
video.cc:(.text+0x65a): undefined reference to `cv::fastFree(void*)'
/home/jenkins/workspace/B_robotics_bob_robotics_uav_misc/make_common/../libbebop/libbebop.a(video.o): In function `BoBRobotics::Video::Input::readGreyscaleFrame(cv::Mat&)':
video.cc:(.text._ZN11BoBRobotics5Video5Input18readGreyscaleFrameERN2cv3MatE[_ZN11BoBRobotics5Video5Input18readGreyscaleFrameERN2cv3MatE]+0x80): undefined reference to `cv::Mat::create(int, int const*, int)'
video.cc:(.text._ZN11BoBRobotics5Video5Input18readGreyscaleFrameERN2cv3MatE[_ZN11BoBRobotics5Video5Input18readGreyscaleFrameERN2cv3MatE]+0xcb): undefined reference to `cv::cvtColor(cv::_InputArray const&, cv::_OutputArray const&, int, int)'
/home/jenkins/workspace/B_robotics_bob_robotics_uav_misc/make_common/../libbebop/libbebop.a(video.o): In function `cv::Mat::~Mat()':
video.cc:(.text._ZN2cv3MatD2Ev[_ZN2cv3MatD5Ev]+0x89): undefined reference to `cv::Mat::deallocate()'
video.cc:(.text._ZN2cv3MatD2Ev[_ZN2cv3MatD5Ev]+0x72): undefined reference to `cv::fastFree(void*)'
collect2: error: ld returned 1 exit status
Makefile:18: recipe for target 'bebop_relative_move_test' failed
make: *** [bebop_relative_move_test] Error 1
```

Any ideas? I thought it might be something to do with the order we're passing the linking arguments to gcc, but OpenCV is earlier in the arguments than the Bebop stuff.

**Comment by alexdewar** [comment]
Weirdly, ``bebop_test``, which uses the same build arguments, compiles fine.

**Comment by neworderofjamie** [comment]
Interesting - my machine, via a forced upgrade to Ubuntu 18, is now using GCC 7.3 and everything! I'll investigate

**Comment by alexdewar** [comment]
Cool, cheers.

**Comment by neworderofjamie** [comment]
Swapping the order the bebop stuff and the opencv stuff are setup in ``bob_robotics.mk`` fixes it but I don't know why

**Comment by alexdewar** [comment]
That's weird -- but if it works then let's do it!

**Comment by neworderofjamie** [comment]
It also works if you add ``Video::Display display(drone.getVideoStream(), true);`` (and relevant includes) to bebop_relative_move_test.cc - something really weird's going on

**Comment by neworderofjamie** [comment]
I think https://stackoverflow.com/questions/36560837/gcc-static-linking-order explains the reason - it's something to do with, in the case of bebop_test (and my modified bebop_relative_move_test), the cc file requires opencv

**Comment by alexdewar** [comment]
That is weird. I've swapped the build args order now and it seems to work though.... Shall we merge?

**Comment by neworderofjamie** [comment]
Sure - at some point I dream of understanding the logic behind the order of gcc flags but now is not that time!

**Comment by alexdewar** [comment]
Cool. Cheers for looking at this!

**Pull request merged**